azure-iot-provisioning-devicesdk/azure/iot/provisioning/devicesdk/transport/mqtt_transport.py

```diff
@@ -0,0 +1,82 @@
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+# --------------------------------------------------------------------------
+
+import uuid
+import logging
+import six.moves.urllib as urllib
+from . import device_provisioning_constant
+from azure.iot.hub.devicesdk.transport.mqtt.mqtt_provider import MQTTProvider
+from .symmetric_key_transport import SymmetricKeyTransport
+
+logger = logging.getLogger(__name__)
+
+
+class MQTTTransport(SymmetricKeyTransport):
+    def __init__(self, provisioning_host, security_provider):
+        SymmetricKeyTransport.__init__(self, security_provider)
+        self._provisioning_host = provisioning_host
+        self._create_mqtt_provider()
+
+    def send_registration_request(self):
+        self._connect()
+
+        rid = uuid.uuid4()
+        topic = "$dps/registrations/PUT/iotdps-register/?$rid=" + str(rid)
+        self._mqtt_provider.publish(topic, " ")
+
+    def query_operation_status(self):
+        pass
+
+    def disconnect(self):
+        logger.info("Calling provider disconnect")
+        self._mqtt_provider.disconnect()
+
+    def _connect(self):
+        logger.info("Calling provider connect")
+        password = self.get_current_sas_token()
+        self._mqtt_provider.connect(password)
+
+    def _create_mqtt_provider(self):
+        client_id = self._security_client.registration_id
+
+        username = (
+            self._security_client.id_scope
+            + "/registrations/"
+            + self._security_client.registration_id
+            + "/api-version="
+            + device_provisioning_constant.API_VERSION
+            + "&ClientVersion="
+            + urllib.parse.quote_plus(device_provisioning_constant.USER_AGENT)
+        )
+
+        hostname = self._provisioning_host
+
+        self._mqtt_provider = MQTTProvider(client_id, hostname, username, ca_cert=None)
+
+        self._mqtt_provider.on_mqtt_connected = self._on_provider_connect_complete
+        self._mqtt_provider.on_mqtt_disconnected = self._on_provider_disconnect_complete
+        self._mqtt_provider.on_mqtt_published = self._on_provider_publish_complete
+        self._mqtt_provider.on_mqtt_subscribed = self._on_provider_subscribe_complete
+        self._mqtt_provider.on_mqtt_message_received = self.on_provider_message_received_callback
+
+    def _on_provider_connect_complete(self):
+        logger.info("_on_provider_connect_complete")
+        topic = "$dps/registrations/res/#"
+        self._mqtt_provider.subscribe(topic, 1)
+
+    def _on_provider_disconnect_complete(self):
+        logger.info("_on_provider_disconnect_complete")
+
+    def _on_provider_publish_complete(self, mid):
+        logger.info("_on_provider_publish_complete")
+
+    def _on_provider_subscribe_complete(self, mid):
+        logger.info("_on_provider_subscribe_complete")
+
+    def on_provider_message_received_callback(self, topic, payload):
```

Comment:
>def on_provider_message_received_callback [](start = 4, length = 41)

Why is this specific callback/handler the only one without a convention private with underscore?

Also why is it the only one with callback in the title?

(It's better without callback tbh)

Reply:
Also why is it the only one with callback in the title?
 -- got copied from the other transport. it is better without callback

---
In reply to: [270205914](https://github.com/Azure/azure-iot-sdk-python-preview/pull/59#discussion_r270205914) [](ancestors = 270205914)